addons/web_editor/static/src/js/editor/odoo-editor/src/utils/sanitize.js

```diff
@@ -168,28 +172,35 @@ function _sanitizeNodeTree(node, root) {
                 }
             }
         }
+        return node.nextSibling;
+    }
+}
 
-        // Ensure elements which should not contain any content are tagged
-        // contenteditable=false to avoid any hiccup.
-        if (
-            isVoidElement(node) &&
-            node.getAttribute('contenteditable') !== 'false'
-        ) {
-            node.setAttribute('contenteditable', 'false');
-        }
-        if (node.firstChild) {
-            _sanitizeNodeTree(node.firstChild, root);
-        }
-        // Update link URL if label is a new valid link.
-        if (node.nodeName === 'A' && anchor && closestElement(anchor) === node) {
-            const linkLabel = node.textContent;
-            const match = linkLabel.match(URL_REGEX);
-            if (match?.[0] === linkLabel && !node.href.startsWith('mailto:')) {
-                const urlInfo = getUrlsInfosInString(linkLabel)[0];
-                node.setAttribute('href', urlInfo.url);
+/**
+ * Sanitize the given node's tree, in depth-first order.
+ *
+ * @param {Node} node
+ * @param {Element} root
+ */
+function _sanitizeNodeTree(node, root) {
+    if (node && node !== closestElement(node, '[data-oe-protected="true"]')) {
+        const anchor = root.ownerDocument.getSelection()?.anchorNode;
+        const next = _sanitizeNode(node, root);
+        if (node.isConnected) {
+            if (node.firstChild) {
+                _sanitizeNodeTree(node.firstChild, root);
+            }
+            // Update link URL if label is a new valid link.
+            if (node.nodeName === 'A' && anchor && closestElement(anchor) === node) {
+                const linkLabel = node.textContent;
+                const match = linkLabel.match(URL_REGEX);
+                if (match?.[0] === linkLabel && !node.href.startsWith('mailto:')) {
+                    const urlInfo = getUrlsInfosInString(linkLabel)[0];
+                    node.setAttribute('href', urlInfo.url);
+                }
             }
         }
-        node = node.nextSibling;
+        _sanitizeNodeTree(next, root);
```

Comment:
> Also, I don't understand why the link stuff here is in the "tree" version of the function.

It has to happen after having handled all of its children because it messes with the content (I remember we had to make sure of that, there was an actual bug because of this, but I don't remember what it was exactly).

> Do we really need to sanitize the nodes added by `fillEmpty` ?

If these are zero-width spaces, yes, but (1) I don't know if that could happen in this context and (2) if we keep my commit that only removes zws in `cleanForSave` then there's no issue.

> We might solve it by making sanitizeNode to return the sanitized node itself, allowing it to return another node in the very specific case where it was completely replaced.

I like this suggestion actually. It would probably require quite some changes though, but it would be a lot clearer.

I'll look into this, thanks for the detailed feedback.

Reply:
> ```js
> let node = root;
> while (node) {
>     node = sanitizeNode(node);  // The node itself might be replaced during sanitization.
>     // Compute next node to sanitize.
>     if (node.firstChild) {
>         node = node.firstChild;
>     } if (node.nextSibling) {
>         node = node.nextSibling;
>     } else {
>         // Go up the tree to the next ancestor which does have a sibling to go to.
>         while (!node.nextSibling && node !== root) node = node.parentNode;
>         node = node !==root && node.nextSibling;
>     }
> }
> ```

This can be simplified further:
```js
let node = root;
while (node) {
    node = sanitizeNode(node, root); // The node itself might be replaced during sanitization.
    node = node.firstChild || node.nextSibling || ancestors(node, root).find(a => a.nextSibling)?.nextSibling;
}
```